### ISSUE TITLE
Scikit-Learn Pipelines and FeatureUnion

### ISSUE THREAD

**Issue by ratchetwrench** [Featuretools/featuretools#163]
### Bug/Feature Request Description

The requested feature is to be able to use the feature matrix transformation from with a Scikit-Learn pipeline. Something like the below:

```python
from sklearn.pipeline import Pipeline

feature_matrix_enc, features_enc = ft.encode_features(feature_matrix, feature_defs)

pipe = Pipeline([
    ("features", feature_matrix_enc),
    ("feature_selection", feature_selection),
    ("estimator", estimator)
])
```
...then when ```pipe.predict(X_test)``` is called the ```transform()``` method encodes the new data the same as:
```python
feature_matrix = ft.calculate_feature_matrix(saved_features, X_test)
```
So that, at the end of the sklearn pipeline the estimator can be called.

-----
*Issues created here on Github are for bugs or feature requests. For usage questions and questions about errors, please ask on Stack Overflow with the [featuretools](https://stackoverflow.com/questions/tagged/featuretools) tag. Check the [documentation](https://docs.featuretools.com/#get-help) for further guidance on where to ask your question.*

**Comment by kmax12** [comment]
Thanks for the feature request! We love scikit-learn over here, so we'll definitely consider this.

One challenge is that sklearn expects the input feature matrix, `X`, to be passed to the pipeline, while featuretools needs the `EntitySet` (which is multiple tables) and the list of instance ids you want to build features for. Essentially, we have two parameters to pass since we work with data before the feature matrix stage. We have some ideas on how to get this integrated, but that's the challenge we face right now.

**Comment by chappers** [comment]
This can sort of be done with a wrapper, though there are obvious downsides...


```
import featuretools as ft
from sklearn.base import TransformerMixin, BaseEstimator
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

class SimpleFeatureTools(BaseEstimator, TransformerMixin):
    def __init__(self, target_entity=None):
        self.target_entity = target_entity
    
    def fit(self, es):
        feature_matrix_customers, features_defs = ft.dfs(entityset=es,
                                                  target_entity=self.target_entity)
        fm_encoded, features_encoded = ft.encode_features(feature_matrix_customers,
                                                  features_defs)
        self.features_encoded = features_encoded
        return self
    
    def transform(self, es):
        return ft.calculate_feature_matrix(self.features_encoded, es)

es_train = ft.demo.load_mock_customer(return_entityset=True, random_seed=0)
es_test = ft.demo.load_mock_customer(return_entityset=True, random_seed=1)

pipeline = make_pipeline(SimpleFeatureTools("customers"), StandardScaler())
pipeline.fit(es_train)
pipeline.transform(es_test)
```

**Comment by kmax12** [comment]
very cool @chappers!

Two thoughts

* right now if you called `fit` followed by `transform` the feature matrix would get calculated twice. Perhaps it makes more sense to call `dfs` with `features_only=True` in `fit` and then only do the calculations inside of `transform`. This would mean getting rid of the internal encode features step. I think that is okay though, since you can do that in a latter step of the pipeline

* what would be the most sklearn-style way of providing the `cutoff_time` argument? my thought is that it would have to be part of `fit` and `transform` because it changes between training and testing, but not sure if that would comply with sklean api

**Comment by chappers** [comment]
@kmax12 

I think the typical scikit interface is something like:

```
class MyCustomTransfomer(TransformerMixin, BaseEstimator):
    def __init__(self, target_entity=None, cutofftime=??, **kwargs):
        self.target_entity = target_entity
        self.cutofftime = cutofftime
        self.__dict__.update(<whatever this is for kwargs>)
    
    def fit(self, X, y=None):
        self.get_params(blah)
        self.set_params(blah)
        return self

    def transform(self, X):
        self.get_params(blah)
        return X_transformed
```

I think using this pattern won't break whatever is going on...

**Comment by kmax12** [comment]
the problem is the cutoff times will be different will be different for the training set and testing set, so it's hard to put in the init.

**Comment by chappers** [comment]
cutoff time probably needs to be part of the entityset definition to be `sklearn`-ish otherwise it'll have to be `kwargs` in the transform, which although is permissible, may violate peoples expectations of the `transform` method.

```
def transform(self, X, **kwargs):
     # etc...
```

**Comment by kmax12** [comment]
initial implementation has been merged in to master from PR #323. Closing this issue for now. Any feedback on using that new sklearn transformer is appreciated!
